src/server_manager/electron_app/digitalocean_oauth.ts

```diff
@@ -0,0 +1,216 @@
+          const bodyJson: {account: Account} = JSON.parse(body);
+          return resolve(bodyJson.account as Account);
+        });
+  });
+}
+
+export interface OauthSession {
+  // The result of the OAuth session, with the authentication token.
+  result: Promise<string>;
+  // Checks if the OauthSession has been cancelled.
+  isCancelled(): boolean;
+  // Cancels the OAuth session. The result promise is automatically rejected
+  // and isCancelled() will return true.
+  cancel(): void;
+}
+
+// Runs the DigitalOcean oauth flow and returns the access token.
+// See https://developers.digitalocean.com/documentation/oauth/ for the API.
+export function runOauth(): OauthSession {
+  const secret = randomValueHex(16);
+
+  const app = express();
+  const server = http.createServer(app);
+  server.on('close', () => console.log('Oauth server closed'));
+
+  let isCancelled = false;
+  app.use((req, res, next) => {
+    if (isCancelled) {
+      res.status(503).send('Authentication cancelled');
+    } else {
+      next();
+    }
+  });
+
+  app.get('/', (request, response) => {
+    response.send(`<html>
+          <head><title>Authenticating...</title></head>
+          <body>
+              <form id="form" method="POST">
+                  <input id="params" type="hidden" name="params"></input>
+              </form>
+              <script>
+                  let params = new URLSearchParams(location.hash.substr(1));
+                  let form = document.getElementById("form");
+                  let targetUrl = params.get("state");
+                  form.setAttribute("action", targetUrl);
+                  document.getElementById("params").setAttribute("value", params);
+                  form.submit();
+              </script>
+          </body>
+      </html>`);
+  });
+
+  const rejectWrapper = {reject: (error: Error) => {}};
+  const result = new Promise<string>((resolve, reject) => {
+    rejectWrapper.reject = reject;
+    app.post('/', bodyParser.urlencoded({type: '*/*', extended: false}), (request, response) => {
+      server.close();
+
+      const requestSecret = request.query.secret;
+      if (requestSecret !== secret) {
+        response.status(400).send('Authentication failed');
+        reject(new Error(`Expected secret ${secret}. Got ${requestSecret}`));
+        return;
+      }
+      const params = new URLSearchParams(request.body.params);
+      if (params.get('error')) {
+        response.status(400).send('Authentication failed');
+        reject(new Error(`DigitalOcean OAuth error: ${params.get('error_description')}`));
+        return;
+      }
+      const accessToken = params.get('access_token');
+      if (accessToken) {
+        getAccount(accessToken)
+            .then((account) => {
+              if (account.status === 'active') {
+                response.send(`<html><script>window.close()</script></html>`);
+              } else {
+                response.redirect('https://cloud.digitalocean.com');
```

Comment:
What's this experience?  Will it work?

Reply:
This shows the email verification or billing page, depending on what step of the account creation the user is. I tested it and it works.